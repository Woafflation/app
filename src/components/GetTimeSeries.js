import React from 'react'
import { connect } from 'react-redux'
import { compose } from 'recompose'
import isEqual from 'lodash.isequal'
import * as actions from './../actions/types.js'

class GetTimeSeries extends React.Component {
  componentDidMount () {
    this.props.fetchTimeseries({
      price: this.props.price,
      devActivity: this.props.devActivity
    })
  }

  componentDidUpdate (prevProps, prevState) {
    if (!isEqual(this.props.price, prevProps.price)) {
      this.props.fetchTimeseries({
        price: this.props.price,
        devActivity: this.props.devActivity
      })
    }
  }

  render () {
    const { render, timeseries = {} } = this.props
    return render({ timeseries })
  }
}

const mapStateToProps = ({ timeseries }) => ({ timeseries })

const mapDispatchToProps = dispatch => ({
  fetchTimeseries: ({ price, devActivity }) => {
    return dispatch({
      type: actions.TIMESERIES_FETCH,
      payload: { price, devActivity }
    })
  }
})

const enhance = compose(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )
)

export default enhance(GetTimeSeries)
