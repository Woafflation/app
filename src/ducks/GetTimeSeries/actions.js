// timeseries
export const TIMESERIES_FETCH = '[timeseries] FETCH'
export const TIMESERIES_FETCH_SUCCESS = '[timeseries] FETCH_SUCCESS'
export const TIMESERIES_FETCH_FAILED = '[timeseries] FETCH_FAILED'

export const fetchTimeseries = settings => ({
  type: TIMESERIES_FETCH,
  payload: { ...settings }
})
