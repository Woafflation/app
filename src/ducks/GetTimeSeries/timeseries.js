import { DEV_ACTIVITY_QUERY } from './queries/dev_activity_query'
import { MARKET_SEGMENT_QUERY } from './queries/market_segment_query'
import { ETH_SPENT_OVER_TIME_QUERY } from './queries/eth_spent_over_time_query'
import { ETH_SPENT_OVER_TIME_BY_ALL_PROJECTS_QUERY } from './queries/eth_spent_over_time_by_all_projects_query'
import { BURN_RATE_QUERY } from './queries/burn_rate_query'
import { HISTORY_TWITTER_DATA_QUERY } from './queries/history_twitter_data_query'
import { SOCIAL_VOLUME_QUERY } from './queries/social_volume_query'
import { DAILY_ACTIVE_DEPOSITS_QUERY } from './queries/daily_active_deposits_query'
import { SOCIAL_DOMINANCE_QUERY } from './queries/social_dominance_query'
import { PERCENT_OF_TOKEN_SUPPLY_ON_EXCHANGES } from './queries/percent_of_token_supply_on_exchanges_query'
import { TOP_HOLDERS_PERCENT_OF_TOTAL_SUPPLY } from './queries/top_holders_percent_of_total_supply'
import { GAS_USED_QUERY } from './queries/gas_used'
import { PROJECT_TREND_HISTORY_QUERY } from './queries/project_trend_history_query'
import { METRIC_ANOMALIE_QUERY } from './queries/metric_anomaly_query'
import { GET_METRIC } from './queries/get_metric'
import { mergeTimeseriesByKey } from './../../utils/utils'
import { HISTORICAL_BALANCE_QUERY } from '../HistoricalBalance/common/queries'

const getMetricPreTransform = ({ getMetric: { timeseriesData } }) =>
  timeseriesData

const queryBuild = (type, preTransform = getMetricPreTransform) => {
  return {
    query: GET_METRIC(type),
    preTransform
  }
}
/*
   getMetric API available metrics could be seen by querying data with "getAvailableMetrics"
*/
const TIMESERIES = {
  ethSpentOverTime: {
    query: ETH_SPENT_OVER_TIME_QUERY,
    preTransform: ({ ethSpentOverTime: { ethSpentOverTime } }) =>
      ethSpentOverTime
  },
  EthSpentOverTimeByAllProjects: {
    query: ETH_SPENT_OVER_TIME_BY_ALL_PROJECTS_QUERY,
    preTransform: ({ ethSpentOverTimeByAllProjects: ethSpentOverTime }) =>
      ethSpentOverTime
  },
  nvt: {
    query: GET_METRIC('nvt'),
    preTransform: getMetricPreTransform
  },
  nvt_transaction_volume: {
    query: GET_METRIC('nvt_transaction_volume'),
    preTransform: getMetricPreTransform
  },
  burnRate: {
    query: BURN_RATE_QUERY
  },
  realized_value_usd: {
    query: GET_METRIC('realized_value_usd'),
    preTransform: getMetricPreTransform
  },
  historyTwitterData: {
    query: HISTORY_TWITTER_DATA_QUERY
  },
  historicalBalance: {
    query: HISTORICAL_BALANCE_QUERY
  },
  price_usd: {
    query: GET_METRIC('price_usd'),
    preTransform: getMetricPreTransform
  },
  volume_usd: {
    query: GET_METRIC('volume_usd'),
    preTransform: getMetricPreTransform
  },
  marketcap_usd: {
    query: GET_METRIC('marketcap_usd'),
    preTransform: getMetricPreTransform
  },
  devActivity: {
    query: DEV_ACTIVITY_QUERY
  },
  dev_activity: {
    query: GET_METRIC('dev_activity'),
    preTransform: getMetricPreTransform
  },
  socialDominance: {
    query: SOCIAL_DOMINANCE_QUERY
  },
  percentOfTokenSupplyOnExchanges: {
    query: PERCENT_OF_TOKEN_SUPPLY_ON_EXCHANGES,
    preTransform: ({ percentOnExchanges }) => percentOnExchanges
  },
  topHoldersPercentOfTotalSupply: {
    query: TOP_HOLDERS_PERCENT_OF_TOTAL_SUPPLY
  },
  age_destroyed: {
    query: GET_METRIC('age_destroyed'),
    preTransform: getMetricPreTransform
  },
  exchange_balance: {
    query: GET_METRIC('exchange_balance'),
    preTransform: getMetricPreTransform
  },
  daily_active_addresses: {
    query: GET_METRIC('daily_active_addresses'),
    preTransform: getMetricPreTransform
  },
  circulation: {
    query: GET_METRIC('circulation'),
    preTransform: getMetricPreTransform
  },
  dormant_circulation: {
    query: GET_METRIC('dormant_circulation'),
    preTransform: getMetricPreTransform
  },
  mvrv_usd: {
    query: GET_METRIC('mvrv_usd'),
    preTransform: getMetricPreTransform
  },
  dailyActiveDeposits: {
    query: DAILY_ACTIVE_DEPOSITS_QUERY
  },
  velocity: {
    query: GET_METRIC('velocity'),
    preTransform: getMetricPreTransform
  },
  transaction_volume: {
    query: GET_METRIC('transaction_volume'),
    preTransform: getMetricPreTransform
  },
  network_growth: {
    query: GET_METRIC('network_growth'),
    preTransform: getMetricPreTransform
  },
  social_volume_total: queryBuild('social_volume_total'),
  gasUsed: {
    query: GAS_USED_QUERY
  },
  socialVolume: {
    query: SOCIAL_VOLUME_QUERY,
    title: 'Social Volume',
    preTransform: data =>
      mergeTimeseriesByKey({
        key: 'datetime',
        timeseries: Object.values(data),
        mergeData: (longestTSData, timeserieData) => ({
          socialVolume: longestTSData.socialVolume + timeserieData.socialVolume,
          datetime: longestTSData.datetime
        })
      })
  },
  trendPositionHistory: {
    query: PROJECT_TREND_HISTORY_QUERY,
    preTransform: ({ getProjectTrendingHistory: data }) =>
      data.filter(({ position }) => position)
  },
  anomalies: {
    query: METRIC_ANOMALIE_QUERY,
    preTransform: ({ metricAnomaly: anomalies = [] }, metricAnomalyKey) => {
      return anomalies.map(anomaly => ({ ...anomaly, metricAnomalyKey })) || []
    }
  },
  marketSegment: {
    query: MARKET_SEGMENT_QUERY,
    preTransform: ({ devActivity }) => devActivity
  },
  mean_dollar_invested_age: {
    query: GET_METRIC('mean_dollar_invested_age'),
    preTransform: getMetricPreTransform
  },
  price_btc: queryBuild('price_btc'),
  price_eth: queryBuild('price_eth'),
  exchange_inflow: queryBuild('exchange_inflow'),
  exchange_outflow: queryBuild('exchange_outflow'),
  github_activity: queryBuild('github_activity'),
  community_messages_count_discord: queryBuild(
    'community_messages_count_discord'
  ),
  community_messages_count_telegram: queryBuild(
    'community_messages_count_telegram'
  ),
  community_messages_count_total: queryBuild('community_messages_count_total'),
  social_dominance_discord: queryBuild('social_dominance_discord'),
  social_dominance_professional_traders_chat: queryBuild(
    'social_dominance_professional_traders_chat'
  ),
  social_dominance_reddit: queryBuild('social_dominance_reddit'),
  social_dominance_total: queryBuild('social_dominance_total'),
  social_dominance_telegram: queryBuild('social_dominance_telegram'),
  social_volume_professional_traders_chat: queryBuild(
    'social_volume_professional_traders_chat'
  ),
  social_volume_reddit: queryBuild('social_volume_reddit'),
  social_volume_discord: queryBuild('social_volume_discord'),
  social_volume_telegram: queryBuild('social_volume_telegram'),
  social_volume_twitter: queryBuild('social_volume_twitter'),
  mean_age: queryBuild('mean_age'),

  mvrv_usd_intraday: queryBuild('mvrv_usd_intraday'),
  stock_to_flow: queryBuild('stock_to_flow')
}

export const hasMetric = metric => !!TIMESERIES[metric]
export const getMetricQUERY = metric => TIMESERIES[metric].query
export const getPreTransform = (name, metricAnomalyKey, alias = name) => {
  const transform = TIMESERIES[name].preTransform

  return ({ data, ...rest }) => ({
    ...rest,
    data: transform ? { [alias]: transform(data, metricAnomalyKey) } : data,
    __metric: alias
  })
}
