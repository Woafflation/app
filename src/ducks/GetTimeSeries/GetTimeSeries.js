import React from 'react'
import { connect } from 'react-redux'
import { compose } from 'recompose'
import isEqual from 'lodash.isequal'
import { fetchTimeseries } from './actions'

class GetTimeSeries extends React.Component {
  componentDidMount () {
    this.props.fetchTimeseries(getMetrics(this.props))
  }

  componentDidUpdate (prevProps, prevState) {
    if (!isEqual(getMetrics(this.props), getMetrics(prevProps))) {
      this.props.fetchTimeseries(getMetrics(this.props))
    }
  }

  render () {
    const { render, timeseries = {} } = this.props
    return render({ ...timeseries })
  }
}

const getMetrics = props => {
  return Object.keys(props)
    .filter(
      metric =>
        metric !== 'store' &&
        metric !== 'render' &&
        metric !== 'fetchTimeseries' &&
        metric !== 'storeSubscription' &&
        metric !== 'timeseries'
    )
    .reduce((acc, metric) => {
      acc = {
        ...acc,
        [metric]: props[metric]
      }
      return acc
    }, {})
}

const mapStateToProps = ({ timeseries }) => ({ timeseries })

const mapDispatchToProps = dispatch => ({
  fetchTimeseries: metrics => {
    return dispatch(fetchTimeseries(metrics))
  }
})

const enhance = compose(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )
)

export default enhance(GetTimeSeries)
