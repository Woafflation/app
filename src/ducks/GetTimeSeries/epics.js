import moment from 'moment'
import { Observable } from 'rxjs'
import * as actions from './actions'
import { handleErrorAndTriggerAction } from './../../epics/utils'
import {
  mergeTimeseriesByKey,
  getTimeFromFromString
} from './../../utils/utils'
import {
  hasMetric,
  getMetricQUERY,
  getTransforms,
  getSettings
} from './timeseries'

const initialMeta = {
  mergedByDatetime: false
}

const mapDataToTimeserie = ({ data, loading, error }, transforms) => {
  const dataAfterPreTransform =
    transforms && transforms.length > 0
      ? transforms.reduce((acc, val) => {
        const metric = Object.keys(data)[0]
        if (val.predicate(metric)) {
          return val.preTransform(data)
        }
        return data
      }, {})
      : data
  const metric = Object.keys(dataAfterPreTransform)[0]
  const items = !error ? dataAfterPreTransform[metric] : []
  const isEmpty = items && items.length === 0
  return {
    [metric]: {
      items,
      isError: error,
      isEmpty,
      isLoading: loading
    }
  }
}

const mapDataToTimeseries = (data = [], transforms = []) => {
  return data.reduce((acc, val) => {
    return {
      ...acc,
      ...mapDataToTimeserie(val, transforms)
    }
  }, {})
}

const mapDataToMergedTimeserieByDatetime = (data = [], transforms = []) => {
  const timeseriesAsSingleObject = mapDataToTimeseries(data, transforms)
  const timeseries = Object.keys(timeseriesAsSingleObject).map(metric => {
    return timeseriesAsSingleObject[metric].items
  })
  return {
    timeseries: mergeTimeseriesByKey({
      timeseries,
      key: 'datetime'
    }),
    ...Object.keys(timeseriesAsSingleObject).reduce((acc, metric) => {
      const { items, ...rest } = timeseriesAsSingleObject[metric]
      acc[metric] = {
        ...rest
      }
      return acc
    }, {})
  }
}

const fetchTimeseriesEpic = (action$, store, { client }) =>
  action$.ofType(actions.TIMESERIES_FETCH).mergeMap(action => {
    const { meta = initialMeta } = action.payload
    const metrics = Object.keys(action.payload)
      .filter(metric => metric !== 'meta')
      .map(metric => {
        if (!hasMetric(metric)) {
          throw new Error(`
            Unsupported metric yet: "${metric}".
            Add a query for this metric and description.
          `)
        }
        return metric
      })

    const transforms = getTransforms(metrics)
    const queries = metrics.map(metric => {
      const { interval, from = null, to = null, ...rest } = action.payload[
        metric
      ]
      return Observable.fromPromise(
        client.query({
          query: getMetricQUERY(metric),
          variables: {
            interval: interval || '1d',
            to: to || moment().toISOString(),
            from: from || getTimeFromFromString(rest.timeRange),
            ...rest
          }
        })
      )
    })
    return Observable.forkJoin(queries)
      .mergeMap(data => {
        const result = meta.mergedByDatetime
          ? mapDataToMergedTimeserieByDatetime(data, transforms)
          : mapDataToTimeseries(data, transforms)
        const settings = getSettings(metrics)
        return Observable.of({
          type: actions.TIMESERIES_FETCH_SUCCESS,
          payload: {
            settings,
            ...result
          }
        })
      })
      .catch(handleErrorAndTriggerAction(actions.TIMESERIES_FETCH_FAILED))
  })

export default fetchTimeseriesEpic
