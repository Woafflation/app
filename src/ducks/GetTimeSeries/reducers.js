import * as actions from './actions'

export const initialState = {}

export const ERRORS = {
  COMPLEXITY: 'COMPLEXITY',
  ANY: 'ANY'
}

const getErrorType = payload => {
  try {
    const message = payload.networkError.result.errors[0].message
    if (/complex/i.test(message)) {
      return ERRORS.COMPLEXITY
    }
  } catch (e) {
    // pass
  }
  return ERRORS.ANY
}

export default (state = initialState, action) => {
  switch (action.type) {
    case actions.TIMESERIES_FETCH:
      return {
        ...state
      }
    case actions.TIMESERIES_FETCH_SUCCESS:
      return {
        ...state,
        ...action.payload
      }
    case actions.TIMESERIES_FETCH_FAILED:
      return {
        ...initialState,
        isError: true,
        error: { ...action.payload },
        errorType: getErrorType(action.payload)
      }
    default:
      return state
  }
}
