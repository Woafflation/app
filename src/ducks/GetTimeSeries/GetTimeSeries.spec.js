/* eslint-env jest */
import React from 'react'
import toJson from 'enzyme-to-json'
import { mount } from 'enzyme'
import configureStore from 'redux-mock-store'
import GetTimeSeries from './GetTimeSeries'

describe('GetTimeSeries', () => {
  const renderMock = jest.fn(({ timeseries }) => (
    <div id='flag'>{JSON.stringify(timeseries)}</div>
  ))
  const mockStore = configureStore([])
  let store

  beforeEach(() => {
    const initialState = {
      timeseries: {
        price: {
          isLoading: false,
          isError: false,
          items: []
        }
      }
    }
    store = mockStore(initialState)
  })

  const getWrapper = store =>
    mount(
      <GetTimeSeries
        price={{
          timeRange: '6m',
          slug: 'santiment',
          interval: '1d'
        }}
        devActivity={{
          from: '2018-12-01',
          to: '2018-12-10',
          slug: 'santiment',
          interval: '1d',
          transform: 'movingAverage',
          movingAverageIntervalBase: 7
        }}
        store={store}
        render={renderMock}
      />
    )

  it('should return history price list', async () => {
    const wrapper = getWrapper(store)
    expect(wrapper.find(GetTimeSeries).exists()).toBeTruthy()
    expect(store.getActions()).toMatchSnapshot()
  })
})
