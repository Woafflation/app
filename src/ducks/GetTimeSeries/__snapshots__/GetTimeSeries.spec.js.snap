// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GetTimeSeries should return history price list 1`] = `
Array [
  Object {
    "payload": Object {
      "devActivity": Object {
        "from": "2018-12-01",
        "interval": "1d",
        "movingAverageIntervalBase": 7,
        "slug": "santiment",
        "to": "2018-12-10",
        "transform": "movingAverage",
      },
      "price": Object {
        "interval": "1d",
        "slug": "santiment",
        "timeRange": "6m",
      },
    },
    "type": "[timeseries] FETCH",
  },
]
`;
